File: backend/app/schemas/firewall.py
"""防火牆相關 API schemas"""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# ─── 基礎型別 ──────────────────────────────────────────────────────────────────


class PortSpec(BaseModel):
    """端口規格（port=0 表示無端口協定，如 icmp/esp 等）

    三種入站存取模式：
    - domain 有值 → 反向代理（Traefik）
    - external_port 有值 → Port 轉發（haproxy）
    - 兩者皆無 → 僅開放防火牆
    """

    port: int = Field(ge=0, le=65535, description="端口號；0 表示無端口協定")
    protocol: str = Field(default="tcp", description="協定 (tcp/udp/icmp/esp/ah/...)")
    external_port: int | None = Field(
        default=None,
        ge=1,
        le=65535,
        description="外網入站 port（Port 轉發用）",
    )
    domain: str | None = Field(
        default=None,
        max_length=255,
        description="對外網域名稱（反向代理用）",
    )
    enable_https: bool = Field(
        default=True,
        description="反向代理是否啟用 HTTPS（Let's Encrypt）",
    )


# ─── 連線管理 ──────────────────────────────────────────────────────────────────


class ConnectionCreate(BaseModel):
    """建立 VM 間連線（或 VM 到網關，或 Internet 入站）"""

    source_vmid: int | None = Field(description="來源 VM ID；None 代表網關（Internet 入站）")
    target_vmid: int | None = Field(
        default=None, description="目標 VM ID；None 代表網關（上網）"
    )
    ports: list[PortSpec] = Field(description="允許通過的端口列表")
    direction: Literal["one_way", "bidirectional"] = Field(
        default="one_way",
        description="連線方向：one_way（單向）或 bidirectional（雙向）",
    )


class ConnectionDelete(BaseModel):
    """刪除 VM 間連線"""

    source_vmid: int | None = Field(description="來源 VM ID；None 代表網關")
    target_vmid: int | None = Field(
        default=None, description="目標 VM ID；None 代表網關"
    )
    ports: list[PortSpec] | None = Field(
        default=None, description="要刪除的端口；None 代表刪除全部連線"
    )


# ─── 防火牆規則 CRUD ───────────────────────────────────────────────────────────


class FirewallRuleCreate(BaseModel):
    """建立防火牆規則（原始 Proxmox 規則）"""

    type: Literal["in", "out"] = Field(description="規則方向")
    action: Literal["ACCEPT", "DROP", "REJECT"] = Field(description="動作")
    source: str | None = Field(default=None, description="來源 IP/CIDR")
    dest: str | None = Field(default=None, description="目標 IP/CIDR")
    proto: str | None = Field(default=None, description="協定 (tcp/udp/icmp)")
    dport: str | None = Field(default=None, description="目標端口或範圍")
    sport: str | None = Field(default=None, description="來源端口或範圍")
    enable: int = Field(default=1, description="是否啟用 (1=是, 0=否)")
    comment: str | None = Field(default=None, description="備註")


class FirewallRuleUpdate(BaseModel):
    """更新防火牆規則"""

    action: Literal["ACCEPT", "DROP", "REJECT"] | None = None
    source: str | None = None
    dest: str | None = None
    proto: str | None = None
    dport: str | None = None
    sport: str | None = None
    enable: int | None = None
    comment: str | None = None


# ─── 佈局管理 ──────────────────────────────────────────────────────────────────


class LayoutNodeUpdate(BaseModel):
    """更新節點位置"""

    vmid: int | None = Field(default=None, description="VM ID；None 代表 gateway")
    node_type: Literal["vm", "gateway"] = Field(description="節點類型")
    position_x: float = Field(description="X 座標")
    position_y: float = Field(description="Y 座標")


class LayoutUpdate(BaseModel):
    """批次更新圖形佈局"""

    nodes: list[LayoutNodeUpdate]


# ─── 回應 schemas ──────────────────────────────────────────────────────────────


class FirewallRulePublic(BaseModel):
    """防火牆規則（回應）"""

    pos: int
    type: str
    action: str
    source: str | None = None
    dest: str | None = None
    proto: str | None = None
    dport: str | None = None
    sport: str | None = None
    enable: int = 1
    comment: str | None = None
    is_managed: bool = Field(
        default=False,
        description="是否由 Campus Cloud 管理（comment 含 campus-cloud: 前綴）",
    )


class FirewallOptionsPublic(BaseModel):
    """防火牆選項（回應）"""

    enable: bool
    policy_in: str
    policy_out: str


class TopologyNode(BaseModel):
    """拓撲圖中的節點"""

    vmid: int | None = None
    name: str
    node_type: Literal["vm", "gateway"]
    vm_type: Literal["qemu", "lxc"] | None = None
    status: str | None = None
    ip_address: str | None = None
    firewall_enabled: bool = False
    position_x: float = 100.0
    position_y: float = 100.0


class TopologyEdge(BaseModel):
    """拓撲圖中的連線"""

    source_vmid: int | None = None
    target_vmid: int | None = None
    ports: list[PortSpec] = []
    direction: Literal["one_way", "bidirectional"] = "one_way"


class TopologyResponse(BaseModel):
    """完整拓撲資料（節點 + 連線）"""

    nodes: list[TopologyNode]
    edges: list[TopologyEdge]


# ─── NAT 規則 ──────────────────────────────────────────────────────────────────


class NATRulePublic(BaseModel):
    """NAT 端口轉發規則（回應）"""

    id: uuid.UUID
    ssh_host: str
    vmid: int
    vm_ip: str
    external_port: int
    internal_port: int
    protocol: str
    created_at: datetime


class ReverseProxyRulePublic(BaseModel):
    """反向代理規則（回應）"""

    id: uuid.UUID
    vmid: int
    vm_ip: str
    domain: str
    zone_id: str | None = None
    internal_port: int
    enable_https: bool
    dns_provider: str
    created_at: datetime


__all__ = [
    "PortSpec",
    "ConnectionCreate",
    "ConnectionDelete",
    "FirewallRuleCreate",
    "FirewallRuleUpdate",
    "LayoutNodeUpdate",
    "LayoutUpdate",
    "FirewallRulePublic",
    "FirewallOptionsPublic",
    "TopologyNode",
    "TopologyEdge",
    "TopologyResponse",
    "NATRulePublic",
    "ReverseProxyRulePublic",
]

File: frontend/src/services/firewall.ts
/**
 * 防火牆 API 服務
 *
 * 此檔案提供防火牆 API 的手動封裝，
 * 待後端 API 穩定後可透過 generate-client.sh 自動生成並取代。
 */

import type { CancelablePromise } from "@/client"
import { OpenAPI } from "@/client"
import { request as __request } from "@/client/core/request"

// ─── 型別定義 ──────────────────────────────────────────────────────────────────

export type PortSpec = {
  port: number
  protocol: string
  external_port?: number | null
  domain?: string | null
  enable_https?: boolean
}

export type ReverseProxyRulePublic = {
  id: string
  vmid: number
  vm_ip: string
  domain: string
  internal_port: number
  enable_https: boolean
  dns_provider: string
  created_at: string
}

export type NATRulePublic = {
  id: string
  ssh_host: string
  vmid: number
  vm_ip: string
  external_port: number
  internal_port: number
  protocol: string
  created_at: string
}

export type TopologyNode = {
  vmid: number | null
  name: string
  node_type: "vm" | "gateway"
  vm_type: "qemu" | "lxc" | null
  status: string | null
  ip_address: string | null
  firewall_enabled: boolean
  position_x: number
  position_y: number
}

export type TopologyEdge = {
  source_vmid: number | null
  target_vmid: number | null
  ports: PortSpec[]
  direction: "one_way" | "bidirectional"
}

export type TopologyResponse = {
  nodes: TopologyNode[]
  edges: TopologyEdge[]
}

export type FirewallRulePublic = {
  pos: number
  type: string
  action: string
  source: string | null
  dest: string | null
  proto: string | null
  dport: string | null
  sport: string | null
  enable: number
  comment: string | null
  is_managed: boolean
}

export type FirewallOptionsPublic = {
  enable: boolean
  policy_in: string
  policy_out: string
}

export type LayoutNodeUpdate = {
  vmid: number | null
  node_type: "vm" | "gateway"
  position_x: number
  position_y: number
}

// ─── FirewallService ────────────────────────────────────────────────────────────

export const FirewallService = {
  /** 取得拓撲資料（節點 + 連線） */
  getFirewallTopology(): CancelablePromise<TopologyResponse> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/firewall/topology",
      errors: { 422: "Validation Error" },
    })
  },

  /** 取得佈局資料 */
  getFirewallLayout(): CancelablePromise<LayoutNodeUpdate[]> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/firewall/layout",
      errors: { 422: "Validation Error" },
    })
  },

  /** 儲存佈局資料 */
  saveFirewallLayout(data: {
    requestBody: { nodes: LayoutNodeUpdate[] }
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "PUT",
      url: "/api/v1/firewall/layout",
      body: data.requestBody,
      mediaType: "application/json",
      errors: { 422: "Validation Error" },
    })
  },

  /** 列出 NAT 端口轉發規則 */
  listNATRules(): CancelablePromise<NATRulePublic[]> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/firewall/nat-rules",
      errors: { 422: "Validation Error" },
    })
  },

  /** 刪除 NAT 端口轉發規則 */
  deleteNATRule(data: {
    ruleId: string
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "DELETE",
      url: "/api/v1/firewall/nat-rules/{rule_id}",
      path: { rule_id: data.ruleId },
      errors: { 404: "Not Found", 422: "Validation Error" },
    })
  },

  /** 同步 NAT 規則到 Gateway VM */
  syncNATRules(): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/firewall/nat-rules/sync",
      errors: { 403: "Forbidden", 422: "Validation Error" },
    })
  },

  /** 列出反向代理規則 */
  listReverseProxyRules(): CancelablePromise<ReverseProxyRulePublic[]> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/firewall/reverse-proxy-rules",
    })
  },

  /** 刪除反向代理規則 */
  deleteReverseProxyRule(data: {
    ruleId: string
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "DELETE",
      url: "/api/v1/firewall/reverse-proxy-rules/{rule_id}",
      path: { rule_id: data.ruleId },
    })
  },

  /** 同步反向代理規則到 Gateway VM */
  syncReverseProxyRules(): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/firewall/reverse-proxy-rules/sync",
      errors: { 403: "Forbidden" },
    })
  },

  /** 建立連線 */
  createFirewallConnection(data: {
    requestBody: {
      source_vmid: number | null
      target_vmid: number | null
      ports: PortSpec[]
      direction: string
    }
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/firewall/connections",
      body: data.requestBody,
      mediaType: "application/json",
      errors: { 400: "Bad Request", 422: "Validation Error" },
    })
  },

  /** 刪除連線 */
  deleteFirewallConnection(data: {
    requestBody: {
      source_vmid: number
      target_vmid: number | null
      ports?: Array<{ port: number; protocol: string }> | null
    }
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "DELETE",
      url: "/api/v1/firewall/connections",
      body: data.requestBody,
      mediaType: "application/json",
      errors: { 422: "Validation Error" },
    })
  },

  /** 列出 VM 防火牆規則 */
  listFirewallRules(data: {
    vmid: number
  }): CancelablePromise<FirewallRulePublic[]> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/firewall/{vmid}/rules",
      path: { vmid: data.vmid },
      errors: { 422: "Validation Error" },
    })
  },

  /** 取得 VM 防火牆選項 */
  getFirewallOptions(data: {
    vmid: number
  }): CancelablePromise<FirewallOptionsPublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/firewall/{vmid}/options",
      path: { vmid: data.vmid },
      errors: { 422: "Validation Error" },
    })
  },

  /** 建立防火牆規則 */
  createFirewallRule(data: {
    vmid: number
    requestBody: {
      type: "in" | "out"
      action: "ACCEPT" | "DROP" | "REJECT"
      source?: string
      dest?: string
      proto?: string
      dport?: string
      enable?: number
      comment?: string
    }
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/firewall/{vmid}/rules",
      path: { vmid: data.vmid },
      body: data.requestBody,
      mediaType: "application/json",
      errors: { 422: "Validation Error" },
    })
  },

  /** 刪除防火牆規則 */
  deleteFirewallRule(data: {
    vmid: number
    pos: number
  }): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "DELETE",
      url: "/api/v1/firewall/{vmid}/rules/{pos}",
      path: { vmid: data.vmid, pos: data.pos },
      errors: { 422: "Validation Error" },
    })
  },
}
